Feature/room db

ethan-223 opened the pull request:
## WaskDatabaseManager

### 싱글톤으로 구현되었습니다 ! - 아래는 동일한 객체가 호출된 것에 대한 증명
<img src="https://user-images.githubusercontent.com/57310034/89708701-5d3f6d80-d9b4-11ea-96f0-3daa39108653.png"/>

> 아직 본격적으로 디비를 사용하지 않아서 어떤 칼럼이 필요할지, 어떤 Dao 메서드가 필요할지 모르겠어서 당장 떠오르는 것만 구현해보았습니다.
개발 중에 필요한 부분이 생기면 추가하여 사용하면 될 것 같습니다 !
---
### 논의점
- 메인 스레드에서 입출력을 허용할 것인가?

### Table
- `replacement_histories`

### Colum
#### `replacement_histories`
- `replacement_histories_id`(INT) : auto update
- `replace_date`(STRING) : YYYY-MM-DD 형태의 교체 날짜 
- `replace_date_month`(INT) : MM에 해당하는 "월" 정보의 정수 값

### Repoitory
- `public void getAll(LoadHistoriesCallback callback)` : 모든 레코드 읽어오기
- `public void getAll(int month, LoadHistoryCallback callback)` : 입력한 월에 해당하는 레코드 읽어오기
- `public void get(String date, GetHistoriesCallback callback)` : 입력한 날짜에 해당하는 레코드 읽어오기
- `public void insert(ReplacementHistoryEntity newReplacementHistory)` : 전체 레코드 입력
- `public void insert(String date)` : 날짜만 입력
- `public void delete(ReplacementHistoryEntity deleteReplacementHistory)` : 특정 레코드 삭제
- `public void delete(String date)` : 특정 날짜를 가진 레코드를 삭제
- `public void deleteAll()` : 모든 레코드 삭제
- `public void update(ReplacementHistoryEntity updatedReplacementHistory)` : 특정 레코드 업데이트

### Callback
```
public interface LoadHistoriesCallback {

    void onHistoriesLoaded(List<ReplacementHistory> histories);

    void onDataNotAvailable();
}

public interface GetHistoryCallback {

    void onTaskLoaded(ReplacementHistory history);

    void onDataNotAvailable();
}
```
### Test
#### code
```
ReplacementHistoryRepository.LoadHistoriesCallback callback = new ReplacementHistoryRepository.LoadHistoriesCallback() {
            @Override
            public void onHistoriesLoaded(List<ReplacementHistory> histories) {
                Log.d(TAG, "onHistoriesLoaded: " + histories);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d(TAG, "onHistoriesLoaded: data is null");
            }
        };

        replacementHistoryRepository = new ReplacementHistoryRepository(this);
        replacementHistoryRepository.deleteAll();
        replacementHistoryRepository.getAll(callback);
        replacementHistoryRepository.insert("2020-08-01");
        replacementHistoryRepository.insert("2020-08-22");
        replacementHistoryRepository.insert("2020-08-14");
        replacementHistoryRepository.insert("2020-08-12");
        replacementHistoryRepository.getAll(callback);
        replacementHistoryRepository.deleteAll();
        replacementHistoryRepository.insert("2020-08-14");
        replacementHistoryRepository.insert("2020-08-12");
        replacementHistoryRepository.delete("2020-08-12");
        replacementHistoryRepository.get("2020-08-14", new ReplacementHistoryRepository.GetHistoryCallback() {

            @Override
            public void onTaskLoaded(ReplacementHistory history) {
                history.setMonthOfReplaceDate(9);
                 replacementHistoryRepository.update(history);
                 replacementHistoryRepository.getAll(callback);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d(TAG, "onDataNotAvailable: data is null");
            }
        });
```
#### result
<img src="https://user-images.githubusercontent.com/57310034/89739769-513fd280-dabe-11ea-9971-4721769ced3c.png"/>

ethan-223 committed:
feat: Room database 구축

Resolves: #15

ethan-223 committed:
feat: 기존 날짜에서 월을 변경할 수 있는 로직 추가

테스트 코드도 함께 추가했습니다.

ethan-223 committed:
feat: ReplacementHistoryRepository 추가

입출력 데이터가 크지 않다는 점, 비동기 처리가 번거로운 점 감안하여 메인 스레드에서 입출력 가능하도록 허용하였습니다.

ethan-223 committed:
refactor: 패키지 구조 변경

ethan-223 committed:
refactor : DateUtils주석 추가

euzl reviewed (approved):
Room DB는 이렇게 사용하는 거군요..!! 배워갑니다. god SGwanwall

#### 메인스레드 입출력 관련 저의 생각
백그라운드에서 처리하는게 좋을 것 같습니다.
(1) 확장기능(마스크 재고, 평균 교체일 등..? )이 생기고 사용자가 1년 넘게 사용한다고 한다면 데이터가 꽤나 커질 수도 있을 것 같습니다.
(2) 메인에서 허용한다면, 딜레이 시간 대비 앱에서 확인하는 작업(00일째 사용중?)이 너무 짧을 것 같습니다.
(3) 만에 하나를 대비해서..?

하지만 일단 지금 상태로 유지하고, 백그라운드 처리는 MVVM과 같이 **icebox에 넣어두는 것**도 좋을 것 같습니다..

euzl commented:
크... 테스트 코드까지...! Yuxy 개발마스터는 달라달라!

<img src="https://user-images.githubusercontent.com/37680108/89716746-94307600-d9ea-11ea-8d05-0dc9cb4f2d1e.png" width="300"> _(재탕)_

jaeryo2357 commented on app/build.gradle:
```diff
@@ -40,4 +40,6 @@ dependencies {
     androidTestImplementation 'androidx.test.ext:junit:1.1.1'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'
 
+    implementation "androidx.room:room-runtime:$room_version"
```
버전 변수 외부화까지 하시다니 좋습니다!

ethan-223 replied:
다른 것들도 이렇게 했으면 해서 카드로 만들어 두었습니다 ! ㅎ

jaeryo2357 commented on app/src/main/java/com/naccoro/wask/replacement/repository/ReplacementHistoryRepository.java:
```diff
@@ -0,0 +1,71 @@
+public class ReplacementHistoryRepository {
+    
+    private static final String TAG = "ReplacementHistoryRepo";
+
```
저희 앱이 큰 Data를 다룰 것 같지는 않지만 Repositoty class에서 
```private List<ReplacementHistoryEntity> chachedReplacementHistory = null```
이라는 한번 가져온 값을 저장해서 또 사용자가 불러올 때 Dao에서 가져오는 것이 아닌 chaced에서 가져와 오버헤드를 줄이면 더 좋을것 같습니다.

zoonun replied:
DAO 호출을 줄일 수 있군요.

ethan-223 replied:
좋은 생각인 것 같습니다 ! 수정했습니다 ㅎ

jaeryo2357 commented on app/src/main/java/com/naccoro/wask/replacement/repository/ReplacementHistoryRepository.java:
```diff
@@ -0,0 +1,71 @@
+        return getAll(-1);
+    }
+
+    public List<ReplacementHistoryEntity> getAll(int month) {
```
room db의 값을 가져오는 부분에서 List가 반환되는 것이 아닌 Callback으로 호출자에게 전달이 되면 좋겠습니다.
google MVP sample 코드에서는 아래와 같은 Callback 함수를 사용합니다.
```java
   interface LoadTasksCallback {

        void onTasksLoaded(List<Task> tasks);

        void onDataNotAvailable();
    }

   interface GetTaskCallback {
 
        void onTaskLoaded(Task task);

        void onDataNotAvailable();
    }

```

ethan-223 replied:
MVP에 적용하기 위해서는 콜백의 형태가 훨씬 유리하겠군요. 수정해보겠습니다. 감사합니다.

ethan-223 replied:
수정하였습니다 !

jaeryo2357 reviewed (approved):
어노테이션과 주석 모두 완벽하고 Room db 로직 코드와 유닛 테스트 클래스까지 너무 깔끔합니다!

zoonun reviewed (approved):
좋은 코드 감사드립니다. 테스트 코드 한번 해볼게요! 다른 분들 코멘트 감사합니다 🦖

ethan-223 committed:
refactor: 캐시 데이터 추가

피드백 반영하였습니다.

ethan-223 committed:
refactor: ReplacementHistoryRepository 콜백으로 이용

ReplacementHistoryEntity --> ReplacementHistory 클래스 이름 변경
캐시를 매번 업데이트 하지 않고, 데이터를 읽을 때 dirty 상태라면 업데이트 하는 것으로 변경
